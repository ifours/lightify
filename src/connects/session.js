import { connect } from 'react-redux';

import { isLoggedIn } from 'selectors';

const mapStateToProps = state => ({
  loggedIn: isLoggedIn(state),
});

const mapDispatchToProps = dispatch => ({
  authorize() {
    dispatch({ type: 'AUTHORIZE_REQUEST' });
  },

  persistToken(payload) {
    dispatch({ type: 'AUTH_TOKEN_RECEIVE', payload });
  }
});

export default function use(Component) {
  return connect(
    mapStateToProps,
    mapDispatchToProps,
  )(Component);
}

use.mapStateToProps = mapStateToProps;
use.mapDispatchToProps = mapDispatchToProps;
