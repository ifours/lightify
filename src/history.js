import createBrowserHistory from 'history/createBrowserHistory';

export default createBrowserHistory({
  basename: process.env.REACT_APP_BASE_HOST_NAME || '',
});
