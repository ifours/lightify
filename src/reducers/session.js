const initialState = {
  accessToken: undefined,
  expiresIn: 0,
  date: 0,
};

export default (state = initialState, action) => {
  const { type, payload } = action;

  switch (type) {
    case 'AUTH_TOKEN_RECEIVE':
      return payload;
    case 'AUTH_TOKEN_EXPIRE':
      return initialState;
    default:
      return state;
  }
};
