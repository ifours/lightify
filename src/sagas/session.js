import { take, call, all, apply } from 'redux-saga/effects';

import Repository from 'services/Repository';
import history from '../history';

export function* authorize() {
  yield take('AUTHORIZE_REQUEST');
  yield call(Repository.authorize);
}

export function* tokenReceive() {
  const { payload } = yield take('AUTH_TOKEN_RECEIVE');
  yield apply(sessionStorage, sessionStorage.setItem, ['session', JSON.stringify(payload)]);
}

export function* tokenExpire() {
  yield take('AUTH_TOKEN_EXPIRE');
  yield apply(sessionStorage, sessionStorage.setItem, ['session', JSON.stringify({})]);
  yield apply(history, history.push, ['/']);
}

export default function* watchSession() {
  yield all([
    authorize(),
    tokenExpire(),
    tokenReceive(),
  ])
}
