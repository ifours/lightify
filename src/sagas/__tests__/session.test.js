import { take, apply } from 'redux-saga/effects';

import history from '../../history';

import { tokenExpire } from '../session';

beforeEach(() => {
  global.sessionStorage = jest.fn();
  global.sessionStorage.setItem = jest.fn();
  global.sessionStorage.getItem = jest.fn();
});

describe('Session saga', () => {
  it('should run tokenExpire', () => {
    let next = {};
    const gen = tokenExpire();

    next = gen.next();
    expect(next.value).toMatchSnapshot();
    expect(next.value).toEqual(take('AUTH_TOKEN_EXPIRE'));

    next = gen.next();
    expect(next.value).toMatchSnapshot();
    expect(next.value).toEqual(apply(sessionStorage, sessionStorage.setItem, ['session', JSON.stringify({})]));

    next = gen.next();
    expect(next.value).toMatchSnapshot();
    expect(next.value).toEqual(apply(history, history.push, ['/']));
  });
});
