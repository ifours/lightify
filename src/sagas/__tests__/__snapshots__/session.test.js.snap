// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Session saga should run authorize 1`] = `
Object {
  "@@redux-saga/IO": true,
  "TAKE": Object {
    "pattern": "AUTHORIZE_REQUEST",
  },
}
`;

exports[`Session saga should run authorize 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`Session saga should run tokenExpire 1`] = `
Object {
  "@@redux-saga/IO": true,
  "TAKE": Object {
    "pattern": "AUTH_TOKEN_EXPIRE",
  },
}
`;

exports[`Session saga should run tokenExpire 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "session",
      "{}",
    ],
    "context": [Function],
    "fn": [Function],
  },
}
`;

exports[`Session saga should run tokenExpire 3`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/",
    ],
    "context": Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    },
    "fn": [Function],
  },
}
`;

exports[`Session saga should run tokenReceive 1`] = `
Object {
  "@@redux-saga/IO": true,
  "TAKE": Object {
    "pattern": "AUTH_TOKEN_RECEIVE",
  },
}
`;

exports[`Session saga should run tokenReceive 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "session",
      "{\\"foo\\":\\"bar\\"}",
    ],
    "context": [Function],
    "fn": [Function],
  },
}
`;

exports[`Session saga should run watcher 1`] = `
Object {
  "@@redux-saga/IO": true,
  "ALL": Array [
     {},
     {},
     {},
  ],
}
`;
